gapis/api/transform2/transform_chain.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/commandGenerator"
+	"github.com/google/gapid/gapis/config"
+)
+
+type transformChain struct {
+	transforms       []Transform
+	out              Writer
+	generator        commandGenerator.CommandGenerator
+	hasBegun         bool
+	hasEnded         bool
+	currentCommandID api.CmdID
+}
+
+// CreateTransformChain creates a transform chain that will run
+// the required transforms on every command coming from commandGenerator
+func CreateTransformChain(generator commandGenerator.CommandGenerator, transforms []Transform, out Writer) *transformChain {
+	return &transformChain{
+		generator:        generator,
+		transforms:       transforms,
+		out:              out,
+		hasBegun:         false,
+		hasEnded:         false,
+		currentCommandID: 0,
+	}
+}
+
+func (chain *transformChain) beginChain(ctx context.Context) error {
+	handleInitialState(chain.out.State())
+
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.BeginTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "Begin Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) endChain(ctx context.Context) error {
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.EndTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "End Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
```

Comment:
Same as BeginChain.

Also these functions are nearly identical. You could implement them using a private shared function that takes a lambda for the diverging behaviour if you wanted.

Reply:
They look very similar but they have some small differences in what they are doing so making it a higher order function does not benefit that much. e.g while begin and end doesn't take cmdID, transform does or, while we give the actual cmdID to mutateAndWrite in transform, we just give 0(or any arbitrary number as they are not part of the cmd list) in begin and end transform